Core/cloudwatch logging

eddiekeller opened the pull request:
<!--
Please make sure to read the Pull Request Guidelines:
https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#pull-requests
-->

#### Description of changes
Issue - https://github.com/aws-amplify/amplify-js/issues/3524

This PR adds an `AWSCloudWatchProvider` and expands on the functionality of the `ConsoleLogger` class to integrate with said provider. Users can use the new functionality to push logs to CloudWatch in their AWS account. If a user adds the `AWSCloudWatchProvider` as a pluggable to the `ConsoleLogger`, then the logger will automatically send their logs to CloudWatch as well as print them to the console.

When configuring amplify via `Amplify.configure()`, the user will just need to add in a `Logging` property that contains a `logGroupName` and `logStreamName` that they wish to write to. If the log group or log stream does not exist, then we will create it for them.

Additionally, users will need to ensure that their accounts have the necessary permissions to interact with CloudWatch. Namely - 

`DescribeLogGroups`
`DescribeLogStreams`
`CreateLogGroup`
`CreateLogStream`
`PutLogEvents`

Example usage - 

```
Amplify.configure({
  Logging: {
    logGroupName: 'plugin-1',
    logStreamName: '1-plugin',
  },
  ...awsExports
});

const logger = new Logger('logBuddy');
Amplify.register(logger);
logger.addPluggable(new AWSCloudWatchProvider());

function App() {
  return (
    <div className="App">
        <p onClick={() => { logger.warn('clicking,, huh?', Date.now()) }}>
          Edit <code>src/App.js</code> and save to reload.
        </p>
    </div>
  );
}
```

#### Description of how you validated changes
Tested via unit testing and local testing with personal AWS account and sample app


#### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are added (https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [n/a] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

svidgen commented:
A few initial questions, presumably just for my own edification:

* Do we need to handle retries for throttling exceptions? Or does the AWS SDK already handle that?
* Is the throttling on [CreateLogStream](https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_CreateLogStream.html) (50 TPS) a per-client limit? A per-log-group limit?
* Is timestamp enough to ensure logs are recorded in order? Or, do we need some sort of locking or debouncing to ensure logs are always sent in the expected order?

```
console.log("oh no! something is broken!");
console.log("... maybe.");
console.log("OH! and remember to pick up dinner on the way home!")
```

eddiekeller commented:
> A few initial questions, presumably just for my own edification:
> 
> * Do we need to handle retries for throttling exceptions? Or does the AWS SDK already handle that?
> * Is the throttling on [CreateLogStream](https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_CreateLogStream.html) (50 TPS) a per-client limit? A per-log-group limit?
> * Is timestamp enough to ensure logs are recorded in order? Or, do we need some sort of locking or debouncing to ensure logs are always sent in the expected order?
> 
> ```
> console.log("oh no! something is broken!");
> console.log("... maybe.");
> console.log("OH! and remember to pick up dinner on the way home!")
> ```

@svidgen Thanks for the questions.

1. The AWS SDK should be handling retries and throttling already. But I can add in some functionality on our end too to try and avoid throttling situations. Shouldn't be too much extra work.
2. It is per client - not per log group.
3. Good question. The order that the logs get sent in doesn't have to do with the timestamp. Each request gets sent with a unique sequence token for the specific log stream. The next sequence token gets generated by CloudWatch when there is a successful log push. Timestamps are moreso used in the CloudWatch client where users can sort by the timestamp of the event. However, I suppose this doesn't necessarily factor in race conditions of asynchronous requests being initially sent in order. The solution for this is probably related to the first question you mentioned. I feel like a simple queue with a bit of management on our end would probably address both points.

eddiekeller commented:
Added in a queue for logging, with support for batched log pushes. Also added safety checks on the size of the log + log batches, a retry in case a log fails due to the improper sequence number being attached to the request, and an interval to monitor the log queue and submit log pushes whenever there are logs present in the queue.

michaeljcalkins commented:
Would this have the functionality to send EMF objects to CloudWatch?
https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Embedded_Metric_Format_Specification.html

Or at least give us the ability to create a function that can create the object to feed the logger?

eddiekeller commented:
> Would this have the functionality to send EMF objects to CloudWatch?

I'm actually not familiar with EMF. Let me take a look at that and see if it would just plug in or if additional functionality would be required.

> On failed messages will this retry, back off, and eventually stop attempting to send a message?

It depends on why the log push failed. If the log push failed because of an incorrect sequence number, then it will back off, attempt to get the correct sequence number, and then perform a retry. It looks like I do need to add a `clearInterval()` at that point though.

eddiekeller commented:
> > Would this have the functionality to send EMF objects to CloudWatch?
> 
> I'm actually not familiar with EMF. Let me take a look at that and see if it would just plug in or if additional functionality would be required.

Looked into it, and it looks like EMF requires a bit of a different format than just sending up logs normally. This is functionality that should be relatively simple to implement, but would probably be best added in a subsequent PR so as to stop the size creep on this one. I am keeping a list of things to add to our logger offering once this gets merged in, and will make additional issues for them so we can track their progress.

jamesaucode reviewed (commented):
Nice work! I left some comments on the code and I have a couple more general questions (before I pull it and try it out on my dev desktop later 😄) :
1. The usage of this would be something like `Logger.addPluggable(new AWSCloudWatchProvider())`?
2. With that, every `logger.x` call will add to the internal buffer of `AWSCloudWatchProvider` and be shipped to CloudWatch every 2 seconds?

eddiekeller commented on packages/core/src/Providers/AWSCloudWatchProvider.ts:
```diff
@@ -0,0 +1,536 @@
+		} catch (err) {
+			logger.error(`error during _safeUploadLogEvents: ${err}`);
+
+			if (RETRY_ERROR_CODES.includes(err.name)) {
```
Yes, the explicit retry here is to handle an edge case race condition where we might attempt to push a log with an improper sequence number. The update I made to utilize a queue should ideally take care of this problem on its own...but I feel it is safest for us to handle an explicit retry on this specific case anyway. I figure it's a low risk-high reward to include it.

eddiekeller commented on packages/core/src/Providers/AWSCloudWatchProvider.ts:
```diff
@@ -0,0 +1,536 @@
+		this._dataTracker.logEvents = [...this._dataTracker.logEvents, ...logs];
+	}
+
+	private async _validateLogGroupExists(
```
Hm. Is the suggestion then a change on the method name just to make the operations performed more clear?

eddiekeller commented:
> Nice work! I left some comments on the code and I have a couple more general questions (before I pull it and try it out on my dev desktop later 😄) :

Thanks for the comments! I will run through them and make the changes highlighted.

> 1. The usage of this would be something like `Logger.addPluggable(new AWSCloudWatchProvider())`?

Yep, you got it. I have a sample code snippet in the PR overview that shows this usage.

> 2. With that, every `logger.x` call will add to the internal buffer of `AWSCloudWatchProvider` and be shipped to CloudWatch every 2 seconds?

Every call of that instance of the logger, yes. We could reduce the time to lower than two seconds, but I chose that as a starting point in order to safely avoid hitting throttle limits on the CloudWatch APIs. But, really, even dropping it down to every 30 seconds or so would be fine. And every log gets published with a timestamp of the event, too.

> 3. Is there a way for customer to specify which level of logs they want to be shipped to Cloudwatch ? For example a user might just want to ship `logger.warn` or `logger.error` but not anything else.

I didn't include that, but that would be simple enough to add. The current behavior of our logger is that the default log level is `WARN`, and unless the user specifies a lower log level when instantiating the logger, then only `WARN` and `ERROR` logs will run through the workflow. But I suppose it is easy enough to add an additional parameter to explicitly set the CloudWatch log level threshold too.

eddiekeller commented on packages/core/src/Providers/AWSCloudWatchProvider.ts:
```diff
@@ -0,0 +1,536 @@
+		while (currentEventIdx < this._dataTracker.logEvents.length) {
+			const currentEvent = this._dataTracker.logEvents[currentEventIdx];
+			const eventSize = currentEvent
+				? Buffer.byteLength(currentEvent.message, 'utf8') +
```
You are correct. Thanks for pointing this out. Looks like the browser way to do this is using `(new TextEncoder().encode(curretEvent.message)).length`

eddiekeller commented on packages/core/src/Providers/AWSCloudWatchProvider.ts:
```diff
@@ -0,0 +1,536 @@
+		 *   ...the log stream does exist but has no logs written to it yet
+		 */
+		try {
+			if (!this._dataTracker.verifiedLogGroup) {
```
Yep, thanks!

eddiekeller commented on packages/core/src/Providers/AWSCloudWatchProvider.ts:
```diff
@@ -0,0 +1,536 @@
+		const client = this._initCloudWatchLogs();
+
+		try {
+			const credentialsOK = await this._ensureCredentials();
```
Yes, good catch.

eddiekeller commented:
> > > 1. Is there a way for customer to specify which level of logs they want to be shipped to Cloudwatch ? For example a user might just want to ship `logger.warn` or `logger.error` but not anything else.
> > 
> > 
> > I didn't include that, but that would be simple enough to add. The current behavior of our logger is that the default log level is `WARN`, and unless the user specifies a lower log level when instantiating the logger, then only `WARN` and `ERROR` logs will run through the workflow. But I suppose it is easy enough to add an additional parameter to explicitly set the CloudWatch log level threshold too.
> 
> I can't envision a scenario where our teams would use an ability like that. If I'm using `logger.*` I'm intentionally sending those logs to CloudWatch otherwise I'd use `console.*`.

I'll leave this as something to investigate in my tracker ticket for future ideas then.

svidgen commented on packages/core/src/Providers/AWSCloudWatchProvider.ts:
```diff
@@ -0,0 +1,536 @@
+		this._dataTracker.logEvents = [...this._dataTracker.logEvents, ...logs];
+	}
+
+	private async _validateLogGroupExists(
```
**Not a blocker** of course, but this name doesn't read to me like a method that will *do* things.

How about something like simplifying the name and adding an arg that communicate intent, like:

```js
_getLogGroup(..., createIfMissing?: boolean = true) {
  ...
}
```

Or something like `_upsertLogGroup(...)`?

svidgen commented on packages/core/src/Providers/AWSCloudWatchProvider.ts:
```diff
@@ -0,0 +1,536 @@
+		}
+	}
+
+	private async _validateLogStreamExists(
```
Ditto some of the above comments. The naming doesn't communicate intent to create (not a blocker for me). I also suspect there are opportunities to DRY things out between this and the other `_validate*` method(s).

svidgen commented on packages/core/src/Providers/AWSCloudWatchProvider.ts:
```diff
@@ -0,0 +1,536 @@
+		}, 2000);
+	}
+
+	private _getDocUploadPermissibility(): boolean {
```
It took me longer to grok this than I care to admit. Could this be accurately called `_canUpload`?

svidgen commented on packages/core/src/Providers/AWSCloudWatchProvider.ts:
```diff
@@ -0,0 +1,536 @@
+		}
+	}
+
+	private _getBufferedBatchOfLogs(): InputLogEvent[] {
```
I suspect this is well-covered indirectly. But, it may be worthwhile to add some explicit test-cases for this, unless I overlooked them.

svidgen reviewed (approved):
I don't think any of my comments were blockers &mdash; Mostly just maintainability and readability paranoia.

eddiekeller commented on packages/core/src/Providers/AWSCloudWatchProvider.ts:
```diff
@@ -0,0 +1,536 @@
+	}
+
+	private _initiateLogPushInterval(): void {
+		if (this._timer) {
+			clearInterval(this._timer);
+		}
+
+		this._timer = setInterval(async () => {
```
I actually just took this timer creation from another one of our Providers.

chemzo commented:
@eddiekeller I see this piece of work has been reverted in #8412 . Is there any plan to incorporate this work again?

alexanderMontague commented:
Hey @eddiekeller 👋 Thanks for re-merging this! Excited to use it. In my testing though I can't seem to actually send any logs to cloudwatch. I am getting denied by a missing `sequenceToken`: 

REQ:
```
{
  logEvents: [{message: "[ERROR] 50:23.354 APP_LOGGER - this is an error!", timestamp: 1627530623355}],
  logGroupName: "TEST_WEB_APP",
  logStreamName: "DEVELOPMENT_LOGS",
  sequenceToken: ""
}
```

RES:
```
{
  message: "1 validation error detected: Value '' at 'sequenceToken' failed to satisfy constraint: Member must have length 
  greater than or equal to 1",
  __type: "InvalidParameterException"
}
```

Through reading up on this and your above comments it seems like a previous request should contain this sequence token, but none do. The initial POSTs for Cognito, `logGroups` and `logStreams` all go through perfectly fine. We only get stuck on the above request ^

I am configuring the logger like you specified above:
```
    // configure amplify
    Amplify.configure({
      Logging: {
        logGroupName: 'TEST_WEB_APP',
        logStreamName: `${ENV}_LOGS`,
        sequenceToken: '123',
      },
      ...awsconfig,
    })

    this.amplifyLogger = new Logger('APP_LOGGER')
    Amplify.register(this.amplifyLogger)
    this.amplifyLogger.addPluggable(new AWSCloudWatchProvider())
```

and even tried throwing a dummy sequence token in but it does not get passed through.

Wondering if I am doing something wrong here, or you can give me some advice.

Thanks!!

szatr commented:
@alexanderMontague I ran into the same issue testing out this functionality with similar config. In my case the root cause seemed to be the fact that `DescribeLogStreams` wasn't returning a `uploadSequenceToken` value because the stream was empty. After adding an event to the stream manually, the `sequenceToken` value was populated correctly and subsequent log calls worked without issue. Perhaps the client can be updated to omit the `sequenceToken` parameter from the `PutLogEventsCommand` when it's empty.

Leejjon commented:
Does this work with SSR on Next.js? Or is this only for logging clientside log messages to CloudWatch?

johnsonfamily1234 commented:
Hey all - this is a great plugin thank you. Question - does this work in all regions?

johnsonfamily1234 commented:
Re: my last comment - I have this up and working in us-east-1 and it works fine. Now trying to implement in ap-southeast-2 with a very similar configuration I'm getting a "AWSCloudWatch - error getting log group - Error: Region is missing" error. I've tried everything I can think of but nothing seems to work. Thoughts anyone?

aaronlna commented:
Is there any work planned to move this provider into documentation? I browsed both the [Amplify TypeDocs](https://aws-amplify.github.io/amplify-js/api/index.html) and the [Amplify Docs](https://docs.amplify.aws/lib/utilities/logger/q/platform/js/) but do not see anything mentioned about this provider. It would be helpful to know what configuration there is without having to dig through code.

aaronlna commented:
@alexanderMontague The issue you're having with `sequenceToken` seems related to https://github.com/aws-amplify/amplify-js/issues/9050. Are you trying to write to an empty log-stream?

batelhi commented:
Can I get an explanation how to add access:
DescribeLogGroups
DescribeLogStreams
CreateLogGroup
CreateLogStream
PutLogEvents
to my amplify app?

Im getting an error:
[ERROR] 29:12.6 AWSCloudWatch - error getting log group - AccessDeniedException: User: arn:aws:sts::630051678415:assumed-role/amplify-intome-dev-232506-authRole/CognitoIdentityCredentials is not authorized to perform: logs:DescribeLogGroups on resource: arn:aws:logs:eu-west-1:630051678415:log-group::log-stream:

jjwallaby commented:
Hi I get this error. 

[11:46:37] [ERROR] 46:37.190 **AWSCloudWatch - error when calling _safeUploadLogEvents in the timer interval - ReferenceError: Can't find variable: TextEncoder**
at node_modules\react-native\Libraries\LogBox\LogBox.js:148:8 in registerError
at node_modules\react-native\Libraries\LogBox\LogBox.js:59:8 in errorImpl
at node_modules\react-native\Libraries\LogBox\LogBox.js:33:4 in console.error
at node_modules\expo\build\environment\react-native-logs.fx.js:27:4 in error
at node_modules\@aws-amplify\core\lib-esm\Logger\ConsoleLogger.js:126:12 in prototype._log
at http://192.168.18.6:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false&minify=false:154365:21 in error
at node_modules\@aws-amplify\core\lib-esm\Providers\AWSCloudWatchProvider.js:582:35 in __generator$argument_1
at http://192.168.18.6:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false&minify=false:161027:26 in step
at node_modules\@aws-sdk\client-cloudwatch-logs\node_modules\@aws-crypto\sha256-js\node_modules\tslib\tslib.js:122:34 in <anonymous>
at node_modules\@aws-sdk\client-cloudwatch-logs\node_modules\@aws-crypto\sha256-js\node_modules\tslib\tslib.js:113:43 in rejected

jjwallaby commented:
I get error
AWSCloudWatch - failure during log push: InvalidSequenceTokenException: The given sequenceToken is invalid. The next expected sequenceToken is: 49623231331024656963658064413629994494964312060433793026

jjwallaby commented:
Full Error
[13:24:15] [ERROR] 24:14.351 AWSCloudWatch - failure during log push: InvalidSequenceTokenException: The given sequenceToken is invalid. The next expected sequenceToken is: 49623231331024656963658064413629994494964312060433793026
at node_modules\react-native\Libraries\LogBox\LogBox.js:148:8 in registerError
at node_modules\react-native\Libraries\LogBox\LogBox.js:59:8 in errorImpl
at node_modules\react-native\Libraries\LogBox\LogBox.js:33:4 in console.error
at node_modules\expo\build\environment\react-native-logs.fx.js:27:4 in error
at node_modules\@aws-amplify\core\lib-esm\Logger\ConsoleLogger.js:126:12 in prototype._log
at http://192.168.18.6:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false&minify=false:154367:21 in error
at node_modules\@aws-amplify\core\lib-esm\Providers\AWSCloudWatchProvider.js:386:36 in __generator$argument_1
at http://192.168.18.6:19000/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&hot=false&minify=false:475064:26 in step
at node_modules\@aws-sdk\client-location\node_modules\@aws-crypto\sha256-js\node_modules\tslib\tslib.js:122:34 in <anonymous>
at node_modules\@aws-sdk\client-location\node_modules\@aws-crypto\sha256-js\node_modules\tslib\tslib.js:113:43 in rejected
at node_modules\react-native\node_modules\promise\setimmediate\core.js:37:13 in tryCallOne
at node_modules\react-native\node_modules\promise\setimmediate\core.js:123:24 in setImmediate$argument_0
at node_modules\react-native\Libraries\Core\Timers\JSTimers.js:130:14 in _callTimer
at node_modules\react-native\Libraries\Core\Timers\JSTimers.js:181:14 in _callImmediatesPass
at node_modules\react-native\Libraries\Core\Timers\JSTimers.js:441:30 in callImmediates
at node_modules\react-native\Libraries\BatchedBridge\MessageQueue.js:387:6 in __callImmediates
at node_modules\react-native\Libraries\BatchedBridge\MessageQueue.js:135:6 in __guard$argument_0
at node_modules\react-native\Libraries\BatchedBridge\MessageQueue.js:364:10 in __guard
at node_modules\react-native\Libraries\BatchedBridge\MessageQueue.js:134:4 in flushedQueue

correadevuk commented:
Hi...Wondering if anyone can help. I'm trying to follow this instructions but it's not working. I've added the required cloudwatch permissions to `amplify-myapp-authRole` and `amplify-myapp-unauthRole`  but when running the application I get loads of errors, for example: `AWSCloudWatch - failure during log group search: Error: No credentials` , `AWSCloudWatch - failure while getting next sequence token: Error: No credentials` etc... Am I adding the necessary permissions into the wrong place?
